service/history/ndc/history_replicator.go

```diff
@@ -274,14 +274,29 @@ func (r *HistoryReplicatorImpl) doApplyEvents(
 		}
 	}()
 
+	mutableState, err := wfContext.LoadMutableState(ctx, r.shardContext)
+
+	// import check
+	versionHistoryPast, _ := r.parseVersionHistoryItems(task.getVersionHistory().Items)
+	if _, isNotFound := err.(*serviceerror.NotFound); len(versionHistoryPast) != 0 && isNotFound {
+		return serviceerrors.NewImportMissingEvent(
+			"Failed to replicate event",
+			task.getNamespaceID().String(),
+			task.getWorkflowID(),
+			task.getRunID(),
+			common.FirstEventID,
+			versionHistoryPast[0].Version,
+			versionHistoryPast[len(versionHistoryPast)-1].EventId,
+			versionHistoryPast[len(versionHistoryPast)-1].Version,
+		)
+	}
```

Comment:
the logic should test if the events to be applied belong to past.

what is asserted here is whether "mutable state is missing && version history contains things from the past"